[Renaming] Update copyright notice: change project name, year, and shorten it

nilsfriess opened the pull request:
This PR updates the copyright notices of all files that previously contained copyright notices using the old name.

The new copyright notice reads:
```c
/*
 * This file is part of AdaptiveCpp (formerly hipSYCL, Open SYCL), an
 * implementation of SYCL and C++ standard parallelism for CPUs and GPUs
 *
 * Copyright (c) 2019-2024 Aksel Alpay and contributors
 * All rights reserved.
 *
 * AdaptiveCpp is released under the BSD 2-Clause "Simplified" License.
 * See file LICENSE in the project root for full license details.
 */
```
The year range depends on the previous copyright notice: if it already contains a range of years, it just updates the end year. If it contains only one year, then it changes it to {old_year} - 2024.

I took the liberty of shortening the whole thing by just referring to the LICENSE file instead of putting it in each file, but I can also copy the complete license back in. Of course I didn't do this by hand, I wrote a Python script to do it, so I can just change the new copyright notice and run it again.

fodinabor commented:
While at it we could include SPDX Identifiers: https://spdx.org/licenses/
`SPDX-License-Identifier:`

al42and commented on include/CL/sycl.hpp:
```diff
@@ -1,28 +1,12 @@
 /*
- * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ * This file is part of AdaptiveCpp (formerly hipSYCL, Open SYCL), an
```
Do you explicitly want to keep the old project names here? The project name is "AdaptiveCpp" now, and the file is part of it. It is not part of hipSYCL anymore (since hipSYCL is no more).

In general, I myself prefer to mention the original name when referring to the project, for clarity. But I am not sure this is helpful in copyright headers.

nilsfriess replied:
I agree with your point. If @illuhad has nothing against it, I remove the hipSYCL/ Open SYCL mentions.

illuhad replied:
I'm fine with removing the mentions of the old project names!

al42and commented on include/CL/sycl.hpp:
```diff
@@ -1,28 +1,12 @@
+ * implementation of SYCL and C++ standard parallelism for CPUs and GPUs
  *
- * Copyright (c) 2018-2020 Aksel Alpay
+ * Copyright (c) 2018-2024 Aksel Alpay and contributors
```
What is the "year" policy going forward?

- Linux Foundation suggests not indicating it at all: https://www.linuxfoundation.org/blog/blog/copyright-notices-in-open-source-software-projects
- FSF suggest listing "the year(s) in which you finished preparing the release": https://www.gnu.org/licenses/gpl-howto.en.html
- GNU suggests listing "each year in which you have made nontrivial changes to the package" (updating all files at once): https://www.gnu.org/prep/maintain/html_node/Copyright-Notices.html

- Curl decided to drop years recently, since they are (allegedly) not required for copyright to hold while being a time sink: https://daniel.haxx.se/blog/2023/01/08/copyright-without-years/
- GROMACS used scripts to update the year every time the file is modified, but a while ago switched to having open-ended "(year-of-file-creation)-" interval (which, now that I read a bit more about it, seems to be not compliant with the US Copyright notice requirement, but that notice is optional and not required to claim the copyright on the work): https://gitlab.com/gromacs/gromacs/-/issues/4322.

illuhad replied:
That's a good point. If it's legal to drop (it seems), I would be fine with following what Linux Foundation recommends, as this seems less work for us.

nilsfriess commented:
I have updated the copyright notice again:
- No year as suggested by the Linux foundation
- Changed `Copyright (c) <year> Aksel Alpay` to `Copyright The AdaptiveCpp contributors` (this is one of the suggestions by the LF mentioned in the post Andrey linked above)
- Added a SPDX identifier

illuhad commented:
> Copyright The AdaptiveCpp contributors

Good change, but to me the capitalization seems a bit off. I think the Linux Foundation link suggests "The AdaptiveCpp Contributors" (with capital C). Or not capitalize the "the".

nilsfriess committed:
Only replace comments at beginning of file

nilsfriess committed:
Fix capitalisation in LICENSE

nilsfriess commented:
Fixed the capitalisation.

And here's the script I used for future reference:
```python
import os
import re

def find_files(root_dirs, extensions):
    """
    Recursively find all files with given extensions in the root directories.
    """
    files = []
    for root_dir in root_dirs:
        for subdir, _, filenames in os.walk(root_dir):
            for filename in filenames:
                if any(filename.endswith(ext) for ext in extensions):
                    files.append(os.path.join(subdir, filename))
    return files

def main(root_dirs, new_comment, dry_run=True, verbose=False):
    # Define the file extensions to search for
    extensions = ['.h', '.hpp', '.cpp', '.cc']
    
    # Find all files with the specified extensions
    files = find_files(root_dirs, extensions)

    skipped_files = []
    files_changed = 0
    for file_path in files:
        # Skip files that have their own copyright notices
        if (file_path == "./include/hipSYCL/common/sbo/small_vector.hpp" or
            file_path == "./include/hipSYCL/sycl/libkernel/detail/fp16/bitcasts.h" or
            file_path == "./include/hipSYCL/sycl/libkernel/spirv/spirv_vars.hpp" or
            file_path == "./include/hipSYCL/sycl/libkernel/spirv/spirv_ops.hpp" or
            file_path == "./include/hipSYCL/sycl/libkernel/spirv/spirv_types.hpp" or
            file_path == "./include/hipSYCL/std/hiplike/bits/basic_string.h" or
            file_path == "./include/hipSYCL/std/hiplike/bits/shared_ptr_base.h"):
           skipped_files.append(file_path)
           continue

        with open(file_path, 'r', encoding='utf-8') as file:
            content = file.read()        

        # Match multiline /* ... */ comments with or without a line starting with `//` after that
        new_content = re.sub(r'^\s*(\/\*[\s\S]*?\*\/(\s*\/\/[^\n]*)?\s*)',
                             new_comment,
                             content,
                             count=1)
        if new_content is content:
            skipped_files.append(file_path)
            continue

        files_changed += 1

        if verbose:
            print("#" * (110 - len(file_path)-1), file_path)
            print(new_content[0:len(new_comment) + 200], "...")
            print("#" * 110)
            print("\n\n")

        if not dry_run:
            with open(file_path, 'w', encoding='utf-8') as file:
                file.write(new_content)

    if verbose:
        print("Skipped files:")
        print("\n".join(skipped_files))
    print(f"Updated {files_changed} files.")

if __name__ == "__main__":
    root_dirs = ['./include', './src']
    new_comment = """/*
 * This file is part of AdaptiveCpp, an implementation of SYCL and C++ standard
 * parallelism for CPUs and GPUs.
 *
 * Copyright The AdaptiveCpp Contributors
 *
 * AdaptiveCpp is released under the BSD 2-Clause "Simplified" License.
 * See file LICENSE in the project root for full license details.
 */
// SPDX-License-Identifier: BSD-2-Clause
"""
    
    main(root_dirs, new_comment, dry_run=False, verbose=True)
```
